 ## import libraries

In [17]:
import pandas as pd
import numpy as np
import os

## import df_ords_prods_merge dataframe.

In [18]:
import pandas as pd

In [19]:
path=r"C:\Users\tjsra\OneDrive\Desktop\INSTACART BASKET ANALYSIS"

In [20]:
df_ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data','Prepared Data', 'ords_prods_merge.pkl'))

## 1. create 'Price_label' column.

### create a subset of first million rows

In [21]:
df = df_ords_prods_merge[:1000000]

In [22]:
df.shape

(1000000, 15)

In [23]:
def price_label(row):

  if row['prices'] <= 5:
    return 'Low-range product'
  elif (row['prices'] > 5) and (row['prices'] <= 15):
    return 'Mid-range product'
  elif row['prices'] > 15:
    return 'High range'
  else: return 'Not enough data'

In [24]:
df['price_range'] = df.apply(price_label, axis=1)

C:\Users\tjsra\AppData\Local\Temp\ipykernel_2304\2848743372.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range'] = df.apply(price_label, axis=1)


In [25]:
df['price_range'].value_counts(dropna = False)

price_range
Mid-range product    655444
Low-range product    328201
High range            16355
Name: count, dtype: int64

In [26]:
df['prices'].max()

24.5

### Use loc() method on df subset to apply price range function

In [27]:
df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'

C:\Users\tjsra\AppData\Local\Temp\ipykernel_2304\1169838859.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'


In [28]:
df.loc[(df['prices'] <= 15) & (df['prices'] > 5), 'price_range_loc'] = 'Mid-range product'

In [29]:
df.loc[df['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [30]:
df['price_range_loc'].value_counts(dropna = False)

price_range_loc
Mid-range product     655444
Low-range product     328201
High-range product     16355
Name: count, dtype: int64

### Use loc() method on entire ord_prods_merge dataframe to apply price range function

In [31]:
df_ords_prods_merge.loc[df_ords_prods_merge['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [32]:
df_ords_prods_merge.loc[(df_ords_prods_merge['prices'] <= 15) & (df_ords_prods_merge['prices'] > 5), 'price_range_loc'] = 'Mid-range product'

In [33]:
df_ords_prods_merge.loc[df_ords_prods_merge['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [34]:
df_ords_prods_merge['price_range_loc'].value_counts(dropna = False)

price_range_loc
Mid-range product     6690367
Low-range product     3100143
High-range product     127743
Name: count, dtype: int64

## Create 'busiest_day' column

In [36]:
df_ords_prods_merge['order_day_of_week'].value_counts(dropna = False)

order_day_of_week
0    1894196
1    1729956
6    1382396
5    1291406
2    1286645
3    1176810
4    1156844
Name: count, dtype: int64

### Use For-loop to determine whether are on 'busiest day' (0=saturday), 'least busy' (4=Wednesday), or 'regular busy' (other days of week)

In [38]:
result = []

for value in df_ords_prods_merge["order_day_of_week"]:
  if value == 0:
    result.append("Busiest day")
  elif value == 4:
    result.append("Least busy")
  else:
    result.append("Regularly busy")

In [39]:
result

['Regularly busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Busiest day',
 'Busiest day',
 'Busiest day',
 'Busiest day',
 'Busiest day',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Busiest day',
 'Least busy',
 'Busiest day',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Bus

### Create new column in ords_prods_merge to view results in context.

In [40]:
df_ords_prods_merge['busiest_day'] = result

In [41]:
df_ords_prods_merge['busiest_day'].value_counts(dropna = False)

busiest_day
Regularly busy    6867213
Busiest day       1894196
Least busy        1156844
Name: count, dtype: int64

## Task: step 2 Update 'busiest day' column to 'busiest days' to identify order_day_of _week as 'busiest days' , 'slowest days' , or 'regularly busy'

### create column 'busiest_days' from 'order_day_of_week' to identify two busiest days(0,1), slowest days(4,3), or regularly busy(all other days).

In [42]:
result_2 = []

for value in df_ords_prods_merge["order_day_of_week"]:
  if value == 0 or value == 1:
    result_2.append("Busiest days")
  elif value == 4 or value == 3:
    result_2.append("Slowest days")
  else:
    result_2.append("Regularly busy")

In [43]:
result_2

['Regularly busy',
 'Regularly busy',
 'Busiest days',
 'Slowest days',
 'Slowest days',
 'Busiest days',
 'Regularly busy',
 'Slowest days',
 'Busiest days',
 'Busiest days',
 'Regularly busy',
 'Slowest days',
 'Slowest days',
 'Regularly busy',
 'Slowest days',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Slowest days',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Regularly busy',
 'Regularly busy',
 'Slowest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Slowest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Slowest days',
 'Regularly busy',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Slowest days',
 'Regularly busy',
 'Busiest days',
 'Regularly busy',
 'Busiest days',
 'Busiest days',

## step 3 Check values of new 'busiest days' column for accuracy with obsevations in markdown format

In [44]:
df_ords_prods_merge['busiest_days'] = result_2

In [45]:
df_ords_prods_merge['busiest_days'].value_counts(dropna = False)

busiest_days
Regularly busy    3960447
Busiest days      3624152
Slowest days      2333654
Name: count, dtype: int64

### check the output of ords_prods_merge with new 'busiest days' column

In [47]:
df_ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   3139998      138            28                  6                 11   
1   1977647      138            30                  6                 17   
2    389851      709             2                  0                 21   
3    652770      764             1                  3                 13   
4   1813452      764             3                  4                 17   

   days_since_last_order  first_order  add_to_cart_order  reordered _merge  \
0                    3.0        False                  5          0   both   
1                   20.0        False                  1          1   both   
2                    6.0        False                 20          0   both   
3                    NaN         True                 10          0   both   
4                    9.0        False                 11          1   both   

     price_range_loc     busiest_day    busiest_days  
0  Mid-range product  Regularly busy  Regularly busy  
1  Mid-range product  Regularly busy  Regularly busy  
2  Mid-range product     Busiest day    Busiest days  
3  Mid-range product  Regularly busy    Slowest days  
4  Mid-range product      Least busy    Slowest days

### The total value counts for the ords_prods_merge dataframe is equal to the total sum of all groupings.The total counts of ords_prods_merge = 9918253 and the total of groupings- (3960447+3624152+2333654)= 9918253.

## step 4 Create new column 'busiest_period_of_day' to identify time periods 'Most orders' , 'Average orders' and ' Fewest orders'

In [48]:
df_ords_prods_merge['order_hour_of_day'].value_counts()

order_hour_of_day
10    847433
11    834539
14    824237
13    819415
15    811711
12    806657
16    772746
9     749590
17    636644
8     530095
18    500081
19    384715
20    296474
7     275188
21    246002
22    194143
23    121172
6      87668
0      66104
1      34377
5      26379
2      20972
4      16846
3      15065
Name: count, dtype: int64

### The value counts listed above will be split into equal third for the following labels: Most orders = 10,11,14,13,15,12,16,9    Fewest orders = 23,6,0,1,5,2,4,3      Average orders = 17,8,18,19,20,7,21,22

### Create  For loop if statement labeling periods of time as 'Most orders', 'Average orders, and 'Fewest orders'

In [49]:
result_3 = []

for value in df_ords_prods_merge["order_hour_of_day"]:
  if value in [10, 11, 14, 15, 13, 12, 16, 9]:
    result_3.append("Most orders")
  elif value in [23, 6, 0, 1, 5, 2, 4, 3]:
    result_3.append("Fewest orders")
  else:
    result_3.append("Average orders")

In [50]:
result_3

['Most orders',
 'Average orders',
 'Average orders',
 'Most orders',
 'Average orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Average orders',
 'Fewest orders',
 'Average orders',
 'Fewest orders',
 'Fewest orders',
 'Fewest orders',
 'Fewest orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Fewest orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Average orders',
 'Most orders',
 'Most ord

### Create new column in 'busiest_period_of_day' in ords_prods_merge

In [51]:
df_ords_prods_merge['busiest_period_of_day'] = result_3

In [52]:
df_ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   3139998      138            28                  6                 11   
1   1977647      138            30                  6                 17   
2    389851      709             2                  0                 21   
3    652770      764             1                  3                 13   
4   1813452      764             3                  4                 17   

   days_since_last_order  first_order  add_to_cart_order  reordered _merge  \
0                    3.0        False                  5          0   both   
1                   20.0        False                  1          1   both   
2                    6.0        False                 20          0   both   
3                    NaN         True                 10          0   both   
4                    9.0        False                 11          1   both   

     price_range_loc     busiest_day    busiest_days busiest_period_of_day  
0  Mid-range product  Regularly busy  Regularly busy           Most orders  
1  Mid-range product  Regularly busy  Regularly busy        Average orders  
2  Mid-range product     Busiest day    Busiest days        Average orders  
3  Mid-range product  Regularly busy    Slowest days           Most orders  
4  Mid-range product      Least busy    Slowest days        Average orders

## step 5 Print frequency of 'busiest_period_of_day' column.

In [53]:
df_ords_prods_merge['busiest_period_of_day'].value_counts(dropna = False)

busiest_period_of_day
Most orders       6466328
Average orders    3063342
Fewest orders      388583
Name: count, dtype: int64

## step 7 Export dataframe as a pickle file to prepared Data folder.

In [54]:
df_ords_prods_merge.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_derived.pkl'))